File: 2025-kotlin/src/main/kotlin/nel/marco/Day6.kt
package nel.marco

import java.math.BigInteger

typealias OperatorAndIndex = Pair<String, Int>


data class Day6(
    val useExample: Boolean = false,
    val useMac: Boolean = true,
) : Day(dayNumber = 6, useExample = useExample, macBook = useMac) {

    /*
    I modified the example and input ot have 0 and + at the end for easier parsing
    xxx xxx  xx xx  0
     xx xx  xxx xx  0
      x xx  xxx xxx 0
    .   .   .   .   +
     */

    override fun answerOne(): String {

        val isExample = readInput[3].contains("*")

        val numbersRow1 = mutableListOf<String>()
        val numbersRow2 = mutableListOf<String>()
        val numbersRow3 = mutableListOf<String>()
        val numbersRow4 = mutableListOf<String>()
        val operator = getOperatorsAndStartingIndex(isExample)


        // get the numbers for the other rows
        populateRows(operator, numbersRow1, numbersRow2, numbersRow3, numbersRow4)

        // removed the extra operator to make my life easier
        operator.removeLast()

        var total = BigInteger.ZERO

        for (x in 0 until operator.size) {
            val newList = getNumbersAsList(x, numbersRow1, numbersRow2, numbersRow3, numbersRow4, isExample)
            total += calculateAnswer(operator[x].first, newList)
        }


        return total.toString()
    }

    private fun getNumbersAsList(
        x: Int,
        numbersRow1: MutableList<String>,
        numbersRow2: MutableList<String>,
        numbersRow3: MutableList<String>,
        numbersRow4: MutableList<String>,
        isExample: Boolean
    ): List<BigInteger> {
        var newList = mutableListOf(numbersRow1[x], numbersRow2[x], numbersRow3[x], numbersRow4[x])
        if (isExample) {
            newList = mutableListOf(numbersRow1[x], numbersRow2[x], numbersRow3[x])
        }
        return newList.map { it.replace(" ", "") }.map { it.toBigInteger() }
    }

    private fun populateRows(
        operator: MutableList<Pair<String, Int>>,
        numbersRow1: MutableList<String>,
        numbersRow2: MutableList<String>,
        numbersRow3: MutableList<String>,
        numbersRow4: MutableList<String>
    ) {
        for (x in 0 until operator.size - 1) {
            val current = operator[x].second
            val next = operator[x + 1].second - 1

            numbersRow1.add(readInput[0].substring(current, next))
            numbersRow2.add(readInput[1].substring(current, next))
            numbersRow3.add(readInput[2].substring(current, next))
            numbersRow4.add(readInput[3].substring(current, next))

        }
    }


    override fun answerTwo(): String {

        val isExample = readInput[3].contains("*")

        val numbersRow1 = mutableListOf<String>()
        val numbersRow2 = mutableListOf<String>()
        val numbersRow3 = mutableListOf<String>()
        val numbersRow4 = mutableListOf<String>()
        val operator = getOperatorsAndStartingIndex(isExample)

        populateRows(operator, numbersRow1, numbersRow2, numbersRow3, numbersRow4)

        // removed the extra operator i added in input
        operator.removeLast()

        var total = BigInteger.ZERO

        for (x in 0..operator.size - 1) {

            val newList = getNumbersAsListButReversed(x, numbersRow1, numbersRow2, numbersRow3, numbersRow4, isExample)
            val answer = calculateAnswer(operator[x].first, newList)

            total += answer
        }


        return total.toString()
    }

    private fun calculateAnswer(
        operator: String,
        newList: List<BigInteger>
    ): BigInteger {
        var answer = BigInteger.ZERO
        if (operator == "*") {
            answer = newList.fold(BigInteger.ONE) { acc, i -> acc.times(i) }
        }

        if (operator == "+") {
            answer = newList.fold(BigInteger.ZERO) { acc, i -> acc.plus(i) }
        }
        return answer
    }

    private fun getNumbersAsListButReversed(
        x: Int,
        numbersRow1: MutableList<String>,
        numbersRow2: MutableList<String>,
        numbersRow3: MutableList<String>,
        numbersRow4: MutableList<String>,
        isExample: Boolean,
    ): List<BigInteger> {
        val newList = mutableListOf<String>()

        for (y in 0..numbersRow1[x].length - 1) {
            val one = numbersRow1[x][y].toString()
            val two = numbersRow2[x][y].toString()
            val three = numbersRow3[x][y].toString()
            val four = if(isExample) "" else numbersRow4[x][y].toString()

            val joinedNumber = listOf(one, two, three, four).joinToString("").trim()
            newList.add(joinedNumber)
        }
        return newList.map { it.toBigInteger() }
    }


    /*
    Because example has only 3 rows and actual input has 4
    
    i need to do special check to grab the correct operator row
     */
    private fun getOperatorsAndStartingIndex(isExample: Boolean): MutableList<OperatorAndIndex> {
        val operators = mutableListOf<OperatorAndIndex>()

        for (index in 0..readInput[0].length - 1) {
            val operator = if (isExample)
                readInput[3][index].toString()
            else
                readInput[4][index].toString()

            if (operator != " ")
                operators.add(operator to index)
        }
        return operators
    }


}

File: rust-2025/src/day6/day6.rs
struct OperatorAndIndex {
    operator: char,
    index: usize,
}

// part 1 took 0.037 ms average (totalRuns=133305; inSeconds=5)
pub(crate) fn part1(result: &[String]) -> String {
    let mut operator = get_operators_and_starting_index(result);

    let mut total: i128 = 0;

    for x in 0..operator.len()-1 {
        let current = operator[x].index;
        let next = operator[x + 1].index;   // <-- FIXED

        let new_list = [
            result[0][current..next].trim().parse::<i128>().unwrap(),
            result[1][current..next].trim().parse::<i128>().unwrap(),
            result[2][current..next].trim().parse::<i128>().unwrap(),
            result[3][current..next].trim().parse::<i128>().unwrap(),
        ];

        total += calculate_answer(operator[x].operator, Vec::from(new_list));
    }

    total.to_string()
}


fn calculate_answer(operator: char, new_list: Vec<i128>) -> i128 {
    let mut answer = 0;
    if (operator == '*') {
        answer = new_list.iter().fold(1, |acc, i| acc * i)
    }

    if (operator == '+') {
        answer = new_list.iter().fold(0, |acc, i| acc + i)
    }
    answer
}

fn get_operators_and_starting_index(result: &[String]) -> Vec<OperatorAndIndex> {
    let mut operators: Vec<OperatorAndIndex> = Vec::new();

    for index in 0..result[0].len() {
        //select teh operator
        let operator: char = result[4][index..index + 1].parse().unwrap();

        if operator != ' ' {
            operators.push(OperatorAndIndex {
                operator: operator,
                index,
            })
        }
    }

     operators
}

//part 2 took 0.034 ms average (totalRuns=146536; inSeconds=5)
pub(crate) fn part2(result: &[String]) -> String {
    let mut row1: Vec<String> = Vec::new();
    let mut row2: Vec<String> = Vec::new();
    let mut row3: Vec<String> = Vec::new();
    let mut row4: Vec<String> = Vec::new();

    let mut operator = get_operators_and_starting_index(&result);

    for x in 0..operator.len() - 1 {
        let current = operator[x].index;
        let next = operator[x + 1].index - 1;

        row1.push(result[0][current..next].to_string());
        row2.push(result[1][current..next].to_string());
        row3.push(result[2][current..next].to_string());
        row4.push(result[3][current..next].to_string());
    }

    // removed the extra operator to make my life easier
    operator.pop();

    let mut total: i128 = 0;

    for x in 0..operator.len() {
        let new_list: Vec<i128> = (0..row1[x].len())
            .map(|y| {
                let one = &row1[x][y..y + 1];
                let two = &row2[x][y..y + 1];
                let three = &row3[x][y..y + 1];
                let four = &row4[x][y..y + 1];

                let new = [one, two, three, four]
                    .join("")
                    .replace(" ", "")
                    .to_string();
                new.parse::<i128>().unwrap()
            })
            .collect();

        total += calculate_answer(operator[x].operator, new_list)
    }

    total.to_string()
}
